core/src/main/java/org/apache/calcite/sql/SqlAsOperator.java

```diff
@@ -102,12 +103,26 @@ public void validateCall(
     // we don't want to validate the identifier.
     final List<SqlNode> operands = call.getOperandList();
     assert operands.size() == 2;
-    assert operands.get(1) instanceof SqlIdentifier;
     operands.get(0).validateExpr(validator, scope);
-    SqlIdentifier id = (SqlIdentifier) operands.get(1);
-    if (!id.isSimple()) {
-      throw validator.newValidationError(id,
-          RESOURCE.aliasMustBeSimpleIdentifier());
+
+    SqlNode asIdentifier = operands.get(1);
+    assert asIdentifier instanceof SqlIdentifier
+        || (asIdentifier instanceof SqlNodeList
+            && validator.getConformance().allowSelectTableFunction());
+
```

Comment:
I think it should throw exception if the conformance does not allow table function  instead of  using `assert`.

Reply:
There is a check for this in the Parser.jj, So this should not happen here. I use the `assert` but not the `exception`.

Reply:
Simple rule of thumb. If you throw an exception (as opposed to using assert) then you must add a test for that exception. If the parser has already checked, and ensured that the condition is impossible, then you will find it impossible to write a test case.

Reply:
Thank @julianhyde . There is a test case in `SqlParserTest` for this exception.